### DISCUSSION TITLE
Element Plus Roadmap

### DISCUSSION THREAD

**Discussion by jw-foss** [element-plus/element-plus#3065] [Announcement]
## 关于 ElementPlus About ElementPlus
ElementPlus 是下一代的 ElementUI，基于 Vue3 的组件库，基本上所有的功能点都是继承自 ElementUI，有 ElementUI 的好的点，也有 ElementUI 坏的点。ElementPlus  的出现，不仅仅是为了继承 ElementUI 好的点，也更是为了摆脱 ElementUI 的历史遗留包袱。

ElementPlus is the next generation of ElementUI built on top of Vue3. All most all the functionalities were inherited from ElementUI. With the pros of ElementUI also the cons of ElementUI. But ElementPlus is more than that because we are proudly calling ourselves **Plus**, we are here to inherit all the good sides of ElementUI and get rid of the bad sides of ElementUI.

### 内容 Table of content 

`O` 表示已经完成，`X` 表示还没有。With `O` means it is available now, `X` means it is not

| 是否完成 progress | 功能概述 name of upgrades |
| -------------------| --------------------------- |
|    O                           |            代码迁移 code migration from Vue2 to  Vue3           |
|    O (Almost done)  | CSS variable |
|  WIP                         | Dark Mode 暗黑模式|
|     O                           |  ES Module  native support   |
|     O                           | [ElSekeleton](https://element-plus.org/#/zh-CN/component/skeleton) |
|          O                      | [ElResult](https://element-plus.org/#/zh-CN/component/result)       |
|           O                     | [ElSpace](https://element-plus.org/#/zh-CN/component/space)       |
|           O                     | [ElDescriptions](https://element-plus.org/#/zh-CN/component/descriptions) |
|            O                    | [ElEmpty](https://element-plus.org/#/zh-CN/component/empty)        |
|             O                   | [ElConfigProvider](https://element-plus.org/#/zh-CN/component/config-provider) | 
|   Half done (keyboard navigation awaits)  |  [列表选择器虚拟化   Virtualized ElSelect](https://element-plus.org/#/zh-CN/component/select-v2)  | 
|  X                              | 虚拟化表格  Virtualized Table |
|  X                              | 文档站重构 Documentation Site Refactor |
|  In parallel               |  修复 Bug Bug fixes |
| WIP                          | So many more |

### beta

大家都知道在去年开始我们已经就在 beta，迟迟没有进入 rc 的阶段，这不仅仅是因为 Vue 本身的 API 那时候还不稳定，我们也经历过很多因为破坏性变动导致的问题。所以我们（特别是我）也很害怕做破坏性变动，这样会让 ElementPlus 给人感觉不稳定，但由于很多方面的原因，ElementPlus 只能暂时保持在 beta 的状态。

Since almost everyone knows that we are at beta since last year, and it took us long enough and we are still in beta, this is not only because of the constantly breaking changes from Vue itself but also because we have some unspeakable reasons. And the team (especially me) is afraid of making breaking changes because that would make people think that ElementPlus is not stable. But because of some reasons that we could not say, ElementPlus can only be beta for now.

### Stable release

我们预计会在今年的年底前把所有的 API 稳定下来，目前的情况来看一般不会有老的 API 变动了，但是后续还有一个 ElVirtualizedTable 的大重构，当然这个会跟 ElSelectV2 一样，暂时作为一个全新的组件出现，我们将在未来让这两个组件完全替代老的组件，尽量做到 API 保持一致。

We are expecting the first stable release by the end of this year, from now on, I think there would be no breaking changes with the currently existing APIs, but we are planning to make a brand new ElVirtualizeTable component just like ElSelectV2. In the future we will replace the old `select` and `table` with `ElVirtualizedSelect` `ElVirtualizedTable`. We will try our best to make the API stays the same.

## 特别致谢 Special thanks

感谢所有参与到 ElementPlus 的建设中的小伙伴，为 ElementPlus 添砖加瓦，一个好的开源产品绝对离不开社区的支持，我们也非常希望能够收到任何一个人的贡献，贡献不分大小。:pray:

你可以为 ElementPlus 做的贡献 Contributions you can make for ElementPlus：
- 修Bug/做功能 Bug fixes /Feature deliver
- 文档更新 Documentation updates 
- 写一个好的 issue. Writing a worth reading issue.

I can't say enough thanks to everyone who participated in building ElementPlus, without the support from the community, no one can make a good open-source product. Everyone's contribution is welcomed, all contributors are equal. :pray:
## A note 

Open-sourcing is never easy, especially when I am participating with ElementPlus, but one thing I would like to assure you guys is that ElementPlus is never and never going to be a product of KPI(OKR).

With love❤️,

ElementPlus Team

**Comment by sxzz** [comment]
We can consider refactoring with `script setup`, it has been officially merged into RFC.

**Reply by Otto-J** [reply]
语法转 setup 列入计划了吗

**Reply by sxzz** [reply]
有计划，但defineProps暂时还有点有问题，考虑等Vue完善好再迁移。

**Comment by OneSeven** [comment]
1. 能不能把drawer的弹出效果给恢复成element-ui的样式啊，现在弹出动画太过缓慢了
2. 关于表格，表格大小变化的过渡的时候会巨卡无比
              场景举例，在表格和侧边导航栏配合使用的时候，当把侧边导航栏收起或者弹出的时候，表格所在元素大小改变，这个时候会特别卡，特别是当行```<slot>```使用很多的时候，三行都会卡。
动态改变表格列的时候，好像不是响应式的，会重载整个表格，表现出不是很平滑
3. 希望增加statistic（:fire:），list和tree-select组件
https://ant.design/components/statistic-cn/
https://ant.design/components/list-cn/
4. drawer在使用嵌套drawer的时候，最底|下drawer希望可以自适应改变大小，有层级感
https://ant.design/components/drawer-cn/
5. switch组件希望可以在按钮内增加文字，文字放在按钮外，占地方，阅读性也不太好
https://ant.design/components/switch-cn/

以上建议😊

**Reply by jw-foss** [reply]
@OneSeven 你的建议我们这边看到了，目前确实有很多的问题，有些我们知道有些我们不知道，我们非常感谢你们这边给到的反馈，这是我们前进的方向和动力，但是迫于人力的有限，有一些问题我们可能暂时没有时间去处理，还请谅解，但是所有的反馈我们这边都会收集和整理。:bow:

**Reply by BoringDays** [reply]
transfer什么时候可以支持表格？message什么时候可以支持全局同时显示最大数量、居中/居左/居右等设置？

**Comment by luwenhegithub** [comment]
希望Virtualized Table的功能快点上:heart_eyes::heart_eyes::heart_eyes:，迫不及待了

**Comment by binvb** [comment]
今天看到同事用的都是beta版本感觉有点不安，如果是vue导致的breaking change，大家都可以理解，最多就是限制下版本，其实大家都这么搞。  
其实如果你们都不安，我们用着beta版本更不安，希望可以推出release版本，谢谢
